### PR TITLE
Extract Datatables

### PR DESCRIPTION
**Pull request by victor-paumier**
- add extract buttons (csv, excel, pdf) on lists
- improve list sections, universities, matching

### CODE REVIEW

File: src/DataTable/Response/SectionListResponse.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+class SectionListResponse extends AbstractDataTableResponse
+{
+    public function configure($entity, array $options): void
+    {
+        if ($entity instanceof Section) {
+            $section = $entity;
+            $userRepo = $this->om->getRepository(User::class);
+
+            $this->values['name'] = $section->getName();
+            $this->values['city'] = (null !== $section->getCity()) ? $section->getCity()->getName() : '';
```
Comment (JeremieSamson) [inline src/DataTable/Response/SectionListResponse.php]:
```suggestion
            $this->values['city'] = $section->getCity() ? $section->getCity()->getName() : '';
```

File: src/DataTable/Response/SectionListResponse.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            $this->values['email'] = $section->getEmail();
+            $this->values['website'] = $section->getWebsite();
+            $this->values['facebook'] = $section->getFacebook();
+            $this->values['twitter'] = $section->getTwitter();
+            $this->values['instagram'] = $section->getInstagram();
+            $this->values['activated'] = ($section->isActivated()) ? $this->translator->trans('yes') : $this->translator->trans('no');
```
Comment (JeremieSamson) [inline src/DataTable/Response/SectionListResponse.php]:
```suggestion
            $this->values['activated'] = $this->translator->trans( $section->isActivated() ? 'yes' : 'no');
```

File: src/DataTable/Response/UserListResponse.php [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    $roles .= ', '.UserRoleEnum::ROLE_TRANSLATOR;
+                }
+            }
+            $roles = ucfirst(strtolower(str_replace('ROLE_', '', $roles)));
+
+            $this->values['firstname'] = $user->getFirstname();
+            $this->values['lastname'] = $user->getLastname();
+            $this->values['email'] = $user->getEmail();
+            $this->values['semester'] = $semester;
+            $this->values['university'] = (null !== $user->getUniversity()) ? $user->getUniversity()->getName() : '';
```
Comment (JeremieSamson) [inline src/DataTable/Response/UserListResponse.php]:
```suggestion
            $this->values['university'] = $user->getUniversity() ? $user->getUniversity()->getName() : '';
```

File: src/DataTable/Response/UserListResponse.php [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            $this->values['languages'] = $languages;
+            $this->values['languagesWanted'] = $languagesWanted;
+            $this->values['local'] = ($user->isLocal()) ? $this->translator->trans('yes') : $this->translator->trans('no');
```
Comment (JeremieSamson) [inline src/DataTable/Response/UserListResponse.php]:
```suggestion
            $this->values['local'] = $this->translator->trans($user->isLocal() ? 'yes' : 'no');
```

File: src/DataTable/Response/UserListResponse.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            $this->values['arrival'] = (null != $user->getArrival()) ? $user->getArrival()->format('d/m/Y') : '';
```
Comment (JeremieSamson) [inline src/DataTable/Response/UserListResponse.php]:
```suggestion
            $this->values['arrival'] = $user->getArrival() ? $user->getArrival()->format('d/m/Y') : '';
```

File: src/DataTable/Response/UserListResponse.php [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            $this->values['createdAt'] = $user->getCreatedAt()->format('d/m/Y');
+            $this->values['nbBuddies'] = $user->getNbBuddies();
+            $this->values['nbBuddiesWanted'] = $user->getNbBuddiesWanted();
+            $this->values['nbReport'] = $user->getNbReport();
+            $this->values['nbWarned'] = $user->getNbWarned();
+            $this->values['nbBuddiesRefused'] = $user->getNbBuddiesRefused();
+            $this->values['nbBuddiesRemoved'] = $user->getNbBuddiesRemoved();
+            $this->values['isProfileCompleted'] = ($user->isReadyForMatching()) ? $this->translator->trans('yes') : $this->translator->trans('no');
```
Comment (JeremieSamson) [inline src/DataTable/Response/UserListResponse.php]:
```suggestion
            $this->values['isProfileCompleted'] = $this->translator->trans($user->isReadyForMatching() ? 'yes' : 'no');
```

File: src/DataTable/Response/UserListResponse.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            $this->values['archive'] = ($user->getArchive()) ? $this->translator->trans('yes') : $this->translator->trans('no');
```
Comment (JeremieSamson) [inline src/DataTable/Response/UserListResponse.php]:
```suggestion
            $this->values['archive'] = $this->translator->trans($user->getArchive() ? 'yes' : 'no');
```

File: src/DataTable/Response/UserListResponse.php [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                               ? $this->router->generate('matching_step2', ['internationalId' => $user->getId()])
+                               : null,
+                'deleteClass' => ($user->getId() !== $user->getId()
+                    && !$this->decisionManager->decide($token, [UserRoleEnum::ROLE_SUPER_ADMIN]))
+                    ? 'user_delete'
+                    : null,
+            ]);
+            $this->values['DT_RowAttr'] = ['data-id' => $user->getId()];
+
+            $this->values['compatibility'] = (\array_key_exists('userToCompare', $options) && $options['userToCompare'] instanceof User)
```
Comment (JeremieSamson) [inline src/DataTable/Response/UserListResponse.php]:
```suggestion
            $this->values['compatibility'] = (\array_key_exists('userToCompare', $options) && $options['userToCompare'] instanceof User)
```suggestion
            $this->values['compatibility'] = \array_key_exists('userToCompare', $options) && $options['userToCompare'] instanceof User
```

File: src/Repository/BuddyRepository.php [hunk 9]
```diff
@@ -22,10 +22,14 @@ public function getListInfo(DataTableModel $dataTableColumns, Section $section =
     {
         $qb = $this->getFilteredBuddies($dataTableColumns->getSearch(), $section);
 
-        $qb
-            ->setFirstResult($dataTableColumns->getFirstResult())
-            ->setMaxResults($dataTableColumns->getMaxResults())
-        ;
+        $qb->setFirstResult($dataTableColumns->getFirstResult());
+
+        if (null !== $dataTableColumns->getMaxResults()) {
```
Comment (JeremieSamson) [inline src/Repository/BuddyRepository.php]:
```suggestion
        if ($dataTableColumns->getMaxResults()) {
```

File: src/Repository/SectionRepository.php [hunk 10]
```diff
@@ -13,14 +15,60 @@ public function __construct(ManagerRegistry $registry)
...
     }
 
-    public function getListInfo(): array
+    public function getListInfo(DataTableModel $dataTableColumns): array
+    {
+        $qb = $this->getFilteredSections($dataTableColumns->getSearch());
+
+        $qb->setFirstResult($dataTableColumns->getFirstResult());
+
+        if (null !== $dataTableColumns->getMaxResults()) {
```
Comment (JeremieSamson) [inline src/Repository/SectionRepository.php]:
```suggestion
        if ($dataTableColumns->getMaxResults()) {
```

File: src/Repository/UniversityRepository.php [hunk 11]
```diff
@@ -15,17 +17,33 @@ public function __construct(ManagerRegistry $registry)
...
-            ->setParameter('section', $section)
-            ->andWhere('section.activated = 1')
-            ->getQuery()
-            ->getResult()
-        ;
+        $qb = $this->getFilteredUniversities($dataTableColumns->getSearch(), $section);
+
+        $qb->setFirstResult($dataTableColumns->getFirstResult());
+
+        if (null !== $dataTableColumns->getMaxResults()) {
```
Comment (JeremieSamson) [inline src/Repository/UniversityRepository.php]:
```suggestion
        if ($dataTableColumns->getMaxResults()) {
```

File: src/Repository/UniversityRepository.php [hunk 12]
```diff
@@ -64,4 +82,43 @@ public function getAllUsed(Section $section = null): array
...
     }
+
+    private function getFilteredUniversities(array $search = [], Section $section = null): QueryBuilder
+    {
+        $qb = $this->createQueryBuilder('u')
+            ->innerJoin('u.section', 's')
+            ->innerJoin('s.city', 'c')
+        ;
+
+        if (null !== $section) {
```
Comment (JeremieSamson) [inline src/Repository/UniversityRepository.php]:
```suggestion
        if ($section) {
```

File: src/Repository/UserRepository.php [hunk 13]
```diff
@@ -26,62 +26,60 @@ public function getListInfo(DataTableModel $dataTableColumns, Section $section =
 
         $qb
             ->setFirstResult($dataTableColumns->getFirstResult())
-            ->setMaxResults($dataTableColumns->getMaxResults())
             ->groupBy('u.id')
         ;
 
+        if (null !== $dataTableColumns->getMaxResults()) {
```
Comment (JeremieSamson) [inline src/Repository/UserRepository.php]:
```suggestion
        if ($dataTableColumns->getMaxResults()) {
```

File: src/Controller/UniversityController.php [hunk 14]
```diff
@@ -23,15 +29,41 @@ class UniversityController extends BaseController
...
+        ;
+
+        return $this->render('University/index.html.twig', $params);
+    }
 
-        return $this->render('University/index.html.twig', [
-            'universities' => $universities,
-        ]);
+    /**
+     * @Route("/list_universities", name="university_list", options={ "expose" : true }, condition="request.isXmlHttpRequest()")
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/Controller/UniversityController.php]: no limitation here, BCs can manage their universities

[the remark this answers is not in the document]
Comment (JeremieSamson) [reply src/Controller/UniversityController.php]: my bad, I was talking about HTTP Method limitation (only get here)

File: src/DataTable/DataTableModel.php [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+    public function setSearch(array $search = []): void
+    {
+        if (empty($search) && \array_key_exists('columns', $this->parameters)) {
+            $this->search = [];
+            foreach ($this->parameters['columns'] as $column) {
+                $columnName = $column['data'];
+
+                $searchValue = $column['search']['value'];
+                if ('' !== $searchValue) {
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/DataTable/DataTableModel.php]: 0 is catch as empty

File: src/Controller/MatchingController.php [hunk 16]
```diff
@@ -26,11 +35,55 @@ class MatchingController extends BaseController
...
+        DataTableResponseFactory $responseFactory
+    ): JsonResponse {
+        $section = ($this->isGranted(UserRoleEnum::ROLE_ADMIN))
+            ? null
+            : $this->getSection($user);
+
+        $dataTableColumns = new DataTableModel($request);
+        $dataTableColumns->setSearch(array_merge($dataTableColumns->getSearch(), ['local' => 0]));
+        $buddies = $userRepo->getListStudentsNotMatched($dataTableColumns, $section);
+        $response = [
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/Controller/MatchingController.php]: done

File: src/Controller/MatchingController.php [hunk 17]
```diff
@@ -41,13 +94,20 @@ public function step1(UserInterface $user, UserRepository $userRepo): Response
...
         UserRepository $userRepo,
         MatchingHelper $matchingHelper,
         User $international
     ): Response {
         $this->denyAccessUnlessGranted(UserVoterEnum::MATCH, $international);
-        $locals = $userRepo->getLocalStudentsNotMatched($this->getSection($user), $international);
+
+        $dataTableColumns = new DataTableModel($request);
+        $dataTableColumns->setSearch(array_merge($dataTableColumns->getSearch(), ['local' => 0]));
+        $locals = $userRepo->getListStudentsNotMatched(
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/Controller/MatchingController.php]: used for locals & scores

File: src/Controller/SectionController.php [hunk 18]
```diff
@@ -24,13 +29,32 @@ class SectionController extends BaseController
...
+    public function index(CityRepository $cityRepo): Response
     {
         return $this->render('Section/index.html.twig', [
-            'sections' => $sectionRepo->getListInfo(),
+            'cities' => $cityRepo->findBy([], ['name' => 'asc']),
         ]);
     }
 
+    /**
+     * @Route("/list_sections", name="section_list", options={ "expose" : true }, condition="request.isXmlHttpRequest()")
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/Controller/SectionController.php]: done

File: src/Controller/SectionController.php [hunk 19]
```diff
@@ -24,13 +29,32 @@ class SectionController extends BaseController
...
+     */
+    public function getListSections(
+        Request $request,
+        SectionRepository $sectionRepo,
+        DataTableResponseFactory $responseFactory
+    ): JsonResponse {
+        $dataTableColumns = new DataTableModel($request);
+        $sections = $sectionRepo->getListInfo($dataTableColumns);
+        $response = [
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/Controller/SectionController.php]: done

File: src/Controller/UniversityController.php [hunk 20]
```diff
@@ -23,15 +29,41 @@ class UniversityController extends BaseController
...
+        UniversityRepository $universityRepo,
+        DataTableResponseFactory $responseFactory
+    ): JsonResponse {
+        $section = $this->isGranted(UserRoleEnum::ROLE_ADMIN)
+            ? null
+            : $this->getSection($user);
+
+        $dataTableColumns = new DataTableModel($request);
+        $universities = $universityRepo->getListInfo($dataTableColumns, $section);
+        $response = [
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/Controller/UniversityController.php]: done

File: src/Controller/UniversityController.php [hunk 21]
```diff
@@ -23,15 +29,41 @@ class UniversityController extends BaseController
...
+            'draw' => $dataTableColumns->getDraw(),
+            'recordsFiltered' => $universityRepo->getNbUniversities($dataTableColumns->getSearch(), $section),
+            'data' => $responseFactory->getResponse(UniversityListResponse::class, $universities),
+        ];
+
+        return new JSonResponse($response, Response::HTTP_OK);
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/Controller/UniversityController.php]: done

File: src/DataTable/DataTableModel.php [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+
+    public function getSearch(): array
+    {
+        return $this->search;
+    }
+
+    public function setFirstResult(int $firstResult = 0): void
+    {
+        if (null == $firstResult && \array_key_exists('start', $this->parameters)) {
+            $this->firstResult = $this->parameters['start'];
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/DataTable/DataTableModel.php]: done

File: src/DataTable/DataTableModel.php [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+
+    public function getFirstResult(): int
+    {
+        return $this->firstResult;
+    }
+
+    public function setMaxResults(int $maxResults = null): void
+    {
+        if (null == $maxResults && \array_key_exists('length', $this->parameters)) {
+            $this->maxResults = $this->parameters['length'];
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/DataTable/DataTableModel.php]: done

File: src/DataTable/DataTableModel.php [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+
+    public function getMaxResults(): ?int
+    {
+        return $this->maxResults;
+    }
+
+    public function setDraw(int $draw = 0): void
+    {
+        if (null == $draw && \array_key_exists('draw', $this->parameters)) {
+            $this->draw = $this->parameters['draw'];
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/DataTable/DataTableModel.php]: done

File: src/DataTable/Response/BuddyListResponse.php [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+            } elseif (BuddyStatusEnum::STATUS_REFUSED === $buddy->getStatus()) {
+                $this->values['status'] = $this->translator->trans('buddy.confirmed.ko');
+            } else {
+                $this->values['status'] = $this->translator->trans('buddy.confirmed.none');
+            }
+            $this->values['section'] = $mentor->getSection()->getName();
+            $this->values['createdAt'] = $buddy->getCreatedAt()->format('d/m/Y');
+            $this->values['mentor'] = $mentor->getFirstName().' '.$mentor->getLastName().'<br>'.$mentor->getEmail();
+            $this->values['mentee'] = $mentee->getFirstName().' '.$mentee->getLastName().'<br>'.$mentee->getEmail();
+            $this->values['reasonRefusal'] = ('' != $buddy->getReasonRefusal()) ? $this->translator->trans('buddy.refused.'.$buddy->getReasonRefusal()) : '';
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/DataTable/Response/BuddyListResponse.php]: done

File: src/DataTable/Response/BuddyListResponse.php [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+            $this->values['warnedAt'] = ($buddy->getWarnedAt()) ? $buddy->getWarnedAt()->format('d/m/Y') : '';
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/DataTable/Response/BuddyListResponse.php]: done

File: src/DataTable/Response/BuddyListResponse.php [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+            $this->values['responseAt'] = ($buddy->getResponseAt()) ? $buddy->getResponseAt()->format('d/m/Y') : '';
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/DataTable/Response/BuddyListResponse.php]: done

File: src/DataTable/Response/BuddyListResponse.php [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+            $this->values['archive'] = ($buddy->getArchive()) ? $this->translator->trans('yes') : $this->translator->trans('no');
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/DataTable/Response/BuddyListResponse.php]: done

File: src/DataTable/Response/BuddyListResponse.php [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+            $this->values['actions'] = $this->templating->render('Default/listActions.html.twig', [
+                'urlShow' => $this->router->generate('buddy_show', ['id' => $buddy->getId()]),
+                'reportClass' => ($buddy->getReport()) ? 'buddy_report' : null,
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/DataTable/Response/BuddyListResponse.php]: done

File: src/DataTable/Response/BuddyListResponse.php [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+                'deleteClass' => (!$buddy->getArchive()) ? 'buddy_delete' : null,
```
[the remark this answers is not in the document]
Comment (victor-paumier) [reply src/DataTable/Response/BuddyListResponse.php]: done

File: src/DataTable/Response/SectionListResponse.php (hunk 1 above)
Comment (victor-paumier) [reply src/DataTable/Response/SectionListResponse.php]: done

File: src/DataTable/Response/SectionListResponse.php (hunk 2 above)
Comment (victor-paumier) [reply src/DataTable/Response/SectionListResponse.php]: done

File: src/DataTable/Response/UserListResponse.php (hunk 3 above)
Comment (victor-paumier) [reply src/DataTable/Response/UserListResponse.php]: done

File: src/DataTable/Response/UserListResponse.php (hunk 4 above)
Comment (victor-paumier) [reply src/DataTable/Response/UserListResponse.php]: done

File: src/DataTable/Response/UserListResponse.php (hunk 5 above)
Comment (victor-paumier) [reply src/DataTable/Response/UserListResponse.php]: done

File: src/DataTable/Response/UserListResponse.php (hunk 6 above)
Comment (victor-paumier) [reply src/DataTable/Response/UserListResponse.php]: done

File: src/DataTable/Response/UserListResponse.php (hunk 7 above)
Comment (victor-paumier) [reply src/DataTable/Response/UserListResponse.php]: done

File: src/DataTable/Response/UserListResponse.php (hunk 8 above)
Comment (victor-paumier) [reply src/DataTable/Response/UserListResponse.php]: done

File: src/Repository/BuddyRepository.php (hunk 9 above)
Comment (victor-paumier) [reply src/Repository/BuddyRepository.php]: done

File: src/Repository/SectionRepository.php (hunk 10 above)
Comment (victor-paumier) [reply src/Repository/SectionRepository.php]: done

File: src/Repository/UniversityRepository.php (hunk 11 above)
Comment (victor-paumier) [reply src/Repository/UniversityRepository.php]: done

File: src/Repository/UniversityRepository.php (hunk 12 above)
Comment (victor-paumier) [reply src/Repository/UniversityRepository.php]: done

File: src/Repository/UserRepository.php (hunk 13 above)
Comment (victor-paumier) [reply src/Repository/UserRepository.php]: done

**Pull request merged**